### ISSUE TITLE
handle overlapping memory regions when slicing correctly

### ISSUE THREAD

**Issue by ahmadia** [numba/numba#230]
We don't currently support slices on overlapping memory regions.  This would be nice to have :)

``` python
def overlapping_copy(a):
   a[1:] = a[0:-1]
```

**Comment by pitrou** [comment]
Indeed, we simply copy in ascending order, which in this case would propagate `a[0]` into the whole array. Unfortunately, I don't know how to handle it for the general case (imagine some sophisticated indexing involving several dimensions, some arrays of indices...) short of always copying the operand, which is rather inefficient.

**Comment by seibert** [comment]
I think we would have to implement this by copying the operand, which I assume is how NumPy gets consistent results.

**Comment by damarro3** [comment]
I think I have found a bug related with this issue.

`def test_python(x):
    x[1, 3:6] = x[1, 2:5]
    return x`

`@njit
def test_numba(x):
    x[1, 3:6] = x[1, 2:5]
    return x`

`x_test = np.array([[3, 7, 1, 0, 0, 0, 0, 0], [4, 3, 4, 6, 5, 0, 0, 0]])`

Without @njit decorator, python code returns:

```
[[3 7 1 0 0 0 0 0]
 [4 3 4 4 6 5 0 0]]
```

but with it, the return is not correct:

```
[[3 7 1 0 0 0 0 0]
 [4 3 4 4 4 4 0 0]]
```

**Comment by Jul3k** [comment]
Has this been fixed?
I get the correct result `array([0, 0, 1, 2, 3, 4, 5, 6, 7, 8])` for:

    @njit
    def overlapping_copy(a):
        a[1:] = a[0:-1]
        return a
    
    a = arange(10)

    overlapping_copy(a)

**Comment by stuartarchibald** [comment]
Based on https://github.com/numba/numba/issues/230#issuecomment-478661724, MWR:
```python
from numba import njit
import numpy as np

@njit
def test(x):
    x[1, 3:6] = x[1, 2:5]
    return x

x_test = np.array([[3, 7, 1, 0, 0, 0, 0, 0], [4, 3, 4, 6, 5, 0, 0, 0]])
print(test(x_test.copy()))
print(test.py_func(x_test.copy()))
```
produces the invalid behaviour.
```
[[3 7 1 0 0 0 0 0]
 [4 3 4 4 4 4 0 0]]
[[3 7 1 0 0 0 0 0]
 [4 3 4 4 6 5 0 0]]
```
